### ISSUE TITLE
Different schema for request & response

### ISSUE THREAD

**Issue by timc13** [ihiji/falcon-marshmallow#12]
Great library!

I like the convention of `{method}_schema` to specify different schemas for different HTTP methods.  What about optionally extending that idea to allow different schemas for the request & response of the same method?  Perhaps something like `{method}_{req_or_resp}_schema`.  I can submit a PR if you like.

**Comment by mplanchard** [comment]
Thanks!

I like this idea, and I can definitely see the use case outside of the standard pre- and post- load/dump methods on a schema.

I'd be happy to put it into our development queue and/or try to get it done on evenings/weekends, but I'd also be glad to review a PR. Submitting a PR will almost certainly get things going more quickly, because we're in a bit of a crunch here at the moment in preparation for a trade show in September.

Any thoughts on an implementation strategy? 

At first blush, it might make sense to pass an optional request or response object to `_get_schema()` and on to `_get_method_schema()`, with `_get_method_schema()` doing an `isinstance()` comparison on the object, if it isn't `None`, to determine whether to look for `resp` or `req` in the schema attribute names. I haven't thought it through, though, and am definitely open to other strategies.

I could also see the case for checking more generically for `{req_or_resp}_schema`, but I'm not certain whether it would resolve with higher or lower priority than `{method}_schema`.

**Comment by timc13** [comment]
Hm, I'm not sure I follow you're idea.  Do you mean something like this?:  
```
class MySchema:
    request = MyRequestSchema()
    response = MyResponseSchema()

class MyResource:
    get_schema = MySchemas
```

That would work, but I think it would be simpler to stick with established conventions, instead of introducing a new variant.  We would just have to establish a hierarchy.  The current hierarchy already prefers `{method}_schema` over `schema`.  I think we just introduce a higher priority naming convention, so the order of resolution becomes:
1. `{method}_{request_or_response}_schema`
2. `{method}_schema`
3. `schema`

**Comment by mplanchard** [comment]
Sorry. I was unclear. I agree with your hierarchy. I was thinking that while we're adding req/resp specific schemas, it might be useful to also specify a `{req_or_resp}_schema` generically, like:

```
class MyResource:
   request_schema = SomeSchema()
```

as well as the `{method}_{request_or_response}_schema` that you suggested.

The question then becomes should the resolution order be:

1. `{method}_{req_or_resp}_schema`
2. `{method}_schema`
3. `{req_or_resp}_schema`
4. `schema`

or 

1. `{method}_{req_or_resp}_schema`
2. `{req_or_resp}_schema`
3. `{method}_schema`
4. `schema`

Because neither one seems strongly "correct" to me, perhaps it's best not to include it for the moment, and just allow the specification or request/response specific schemas for a given method.

**Comment by timc13** [comment]
I see...I realize a hierarchy is necessary in any implementation.

Going back to your original idea, I see how the "schema container" implementation could potentially be more flexible and allows the user to optionally declare the hierarchy with the trade off of a more complex API.  

I agree, that the "naming convention" implementation is slightly more inflexible in that we probably can't introduce a `{req_or_resp}_schema` without confusion, but it keeps the API simpler.

I don't feel too strongly either way.  I'll leave it up to you as the maintainer to make a decision! 😄

**Comment by mplanchard** [comment]
I think keeping the API simple is good for now. No sense in complicating things more than is necessary!

So, let's go with just implementing `{method}_{req_or_resp}_schema` for now, with the hierarchy you suggested.

I'll put it into our development queue, but please do open a PR if you've got the time and would like to see it done sooner!

**Comment by mplanchard** [comment]
Resolved by #13

**Comment by mplanchard** [comment]
@timc13 went ahead and released as v 0.2.0, should be available on PyPI. Thanks for the contribution!

**Comment by dlopuch** [comment]
Came here to suggest this very issue!  Happy to see it's already implemented -- thanks! -- needs to be included in the docs though: https://github.com/ihiji/falcon-marshmallow/blame/development/README.rst#L36
